### ISSUE TITLE
Markdown parser produces invalid HTML

### ISSUE THREAD

**Issue by solomax** [vsch/flexmark-java#384]
After upgrading maven-site-plugin to 3.8.2 I found weird output:
```
line [3] Error parsing the model: end tag name </p> must match start tag name <div> from line 3 (position: TEXT seen ...out bd-callout-danger"> It should be run under same user as OM</p>... @3:271)
```

As a result https://issues.apache.org/jira/browse/MSITE-854 has been filed
During investigation Michael Osipov find out the bug should be reported here

could you please take a look?

The markdown being parsed is very simple: https://github.com/apache/openmeetings/blob/master/openmeetings-server/src/site/markdown/InstallMediaServer.md

**Comment by michael-o** [comment]
I am curious what causes the bug.

**Comment by vsch** [comment]
@solomax, can you please tell me what the issue is with the generated HTML. I ran the source through tests on master branch (upcoming 0.60 version), 0.50 and 0.42 and in all cases the result is the same and I cannot see what is incorrect. Note, that → is a tab character changed in tests to visually show tabs for easy identification, it is a `\t` in actual source and HTML.

Source:

```markdown
<!-- 
# Licensed under the Apache License, Version 2.0 (the "License") http://www.apache.org/licenses/LICENSE-2.0 -->

# Media Server Installation

## Install Kurento Media server

<a href="https://doc-kurento.readthedocs.io/en/stable/user/installation.html">Install Kurento Media server</a>
<div class="bd-callout bd-callout-danger">
→It should be run under same user as OM
</div>

## Specify/Install Turn server

<div class="bd-callout bd-callout-info">Optional step</div>
```

Resulting HTML:

```html
<!-- 
# Licensed under the Apache License, Version 2.0 (the "License") http://www.apache.org/licenses/LICENSE-2.0 -->
<h1>Media Server Installation</h1>
<h2>Install Kurento Media server</h2>
<p><a href="https://doc-kurento.readthedocs.io/en/stable/user/installation.html">Install Kurento Media server</a></p>
<div class="bd-callout bd-callout-danger">
→It should be run under same user as OM
</div>
<h2>Specify/Install Turn server</h2>
<div class="bd-callout bd-callout-info">Optional step</div>
```

Maybe based on selected options or some post processing after rendering, the generated HTML is not what I am seeing. However, I don't see it in the library generated HTML.

Can you get the parser options used so I can test the exact options instead of CommonMark defaults.

**Comment by vsch** [comment]
@solomax, @michael-o, it would also help to know the version of the library used so I can have identical conditions for the test.

**Comment by solomax** [comment]
@michael-o can you please provide the options and the exact version being used? (should be easy for you, since you was able to debug it :))

**Comment by michael-o** [comment]
Dependencies: https://maven.apache.org/plugins/maven-site-plugin/dependencies.html, using 0.42.8
Configuration: https://github.com/apache/maven-doxia/blob/doxia-1.9/doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java#L131-L149

**Comment by vsch** [comment]
@solomax, @michael-o, the issue is caused by HTML parser in deep html parse mode not interrupting paragraph blocks on HTML block tags.

The bug is fixed in version 0.42.14, repo for branch 0.42 is updated. Maven central updated and may take some time to show the new version.

A workaround is possible by adding a blank line before the `<div>` tag so that the previous paragraph element is closed so that the HTML block tag will be recognized.

**Comment by michael-o** [comment]
@vsch Very nice! @solomax Can you update the dependency locally and try again?

**Comment by solomax** [comment]
@vsch, @michael-o 
Sorry for delay
Just have checked there is no error with version `0.42.14`
Thanks for quick fix!

**Comment by michael-o** [comment]
@solomax Please another JIRA issue to have flexmark updated with Doxia.

**Comment by hboutemy** [comment]
> The bug is fixed in version 0.42.14, repo for branch 0.42 is updated. Maven central updated and may take some time to show the new version.
0.42 branch: https://github.com/vsch/flexmark-java/commit/a3fcc755bc1c3d7607beff15f70bd714d6a1f16c
0.50 branch: https://github.com/vsch/flexmark-java/commit/40a7fdd0d21c5203c39422cba7cc8b813e1917bc
master branch: https://github.com/vsch/flexmark-java/commit/1f28ad069fcc490d8ae0fe3309238919da56b1cb
